san/impl/params_resolver.cpp

```diff
@@ -0,0 +1,123 @@
+#include <iostream>
+
+#include "san.hpp"
+namespace san {
+
+using namespace ast;
+
+void ParamsValidator::visit(Program* node) {
+    // initial value of a varible could be an expression
+    for (auto var : node->variables) {
+        if (var->type != nullptr) {
+            var->type->accept(*this);
+        }
+    }
+    // length of array could be an expression
```

Comment:
True, but if that expression is a routine call, the expression is no longer a compile-time constant which makes it an invalid array length

Reply:
And what about routines which always return the same value?

Reply:
we do not currently detect that. It could be a optimization, but we decided we don't have time for these manual optimizations and we will just depend on LLVM for now